### ISSUE TITLE
How to work with objects?

### ISSUE THREAD

**Issue by rikakomoe** [LingDong-/wenyan-lang#20]
古有一道法，以一物照一物，謂之曰「射」。 凡變數以「射」為原型者，謂之曰「對象」。 今有程序員者一人，拜讀此作而不知「對象」應作何書，方前來求問。

**Comment by LingDong-** [comment]
Hi there,

Good question! Objects currently share the same interface as Arrays (`列`), because in JavaScript Arrays are also objects. As for compiling to Python, I implemented a custom container class in Python to simulate the same behavior.

```
吾有一列。名之曰「物」。昔之「物」之「「名」」者。今「「甲」」是矣。
var obj = {}; obj.name="a";
```

I know! the above is not intuitive and somewhat confusing. I'm considering about separating Objects from Arrays and give them a good name like `物`.

Another good question is whether or not to bring more sophisticated OOP into wenyan, such as classes, inheritance, constructors, etc. I'll need to think more about that.

Thanks!

**Comment by rikakomoe** [comment]
One more question, how do you identify an object from an array? AFAIC, `吾有一列。名之曰「物」。昔之「物」之「「名」」者。今「「甲」」是矣。` should make `var obj = []; obj.name = 'a'`. Will objects created by this semantics share the prototype `Array`?

**Comment by LingDong-** [comment]
That's correct, you cannot tell Array from objects in my current implementation. But as I mentioned I'm considering adding real objects to the language, keep tuned :)

**Comment by yyx990803** [comment]
Suggestion for object syntax:

```
吾有一物。名之曰「对象」。
夫「对象」者。
  有一數。曰三。名之曰「甲」。
  有一言。曰「「噫吁戲」」。名之曰「乙」。
  有一爻。曰陰。名之曰「丙」。
云云。
書之。
```

Compiles to:

``` js
var DUI4XIANG4 = {
  JIA3: 3,
  YI3: "噫吁戲",
  BING3: false
};
console.log(DUI4XIANG4);
```

**Comment by LingDong-** [comment]
@yyx990803 Nice! I like it very much

**Comment by rikakomoe** [comment]
I think we may need a `delete` syntax like
```
昔「对象」之「甲」者。今不復存矣。
```
compiles to
```javascript
delete DUI4XIANG4.JIA3;
```

BTW, will we consider add a `null` (or even `undefined`) and a `typeof` operator?

**Comment by LingDong-** [comment]
@rikakomoe These suggestions are all great, thanks. Will incorporate.

**Comment by LingDong-** [comment]
Hi everyone!

Object literals are here!
See commits e229e4841a5985fb2b831c37620f3fb2c7b65d81 and 58bee3cfc1843a37a5fd79c4bf0e85c2038e595b.

Inspired by @yyx990803 and @rikakomoe 's suggestions, the basic syntax is like this:

```
吾有一物。名之曰「精衛」。其物如是。
	物之「「名」」者。言曰「「女娃」」。
	物之「「足數」」者。數曰二。
	物之「「喙數」」者。數曰一。
是謂「精衛」之物也。
```

Compiles to 

```
var 精衛={};精衛={"名":"女娃","足數":2,"喙數":1,};
```

Classes are not added yet, but now you can simulate them with "factory" style:

```
吾有一術。名之曰「造物之術」。欲行是術。必先得一言。曰「甲」。乃行是術曰。
	吾有一物。名之曰「乙」。其物如是。
		物之「「丙」」者。言曰「甲」。
		物之「「丁」」者。數曰四。
	是謂「乙」之物也。
	乃得「乙」。
是謂「造物之術」之物也。

施「造物之術」於「「某甲」」。名之曰「戊」。
昔之「戊」之「「丁」」者。今五是矣。
施「JSON.stringify」於「戊」。書之。
```

... and a bit of `JSON.stringify` hack to display the result.

The `delete` syntax:

```
昔之「戊」之「「丁」」者。今不復存矣。
施「JSON.stringify」於「戊」。書之。
```

A new example containing all the above is added to `./examples/obj.wy`. You can try it out on the online IDE too.

There might be some minor bugs with the implementation as it was freshly made, Please let me know if you find any bugs!

Regarding the design, I'm also open to changes if you can convince me ;)

Thank all of you again for the proposal. Let me know what you think. If this one goes down well my next step is to implement import statements. Stay tuned!

**Comment by yyx990803** [comment]
`施「JSON.stringify」於「戊」。` 突然有穿越的感觉，哈哈哈

**Comment by bilcos90** [comment]
class或可称为“象”，取自“以制器者尚其象”

**Comment by bilcos90** [comment]
lamba算子或template function相关的定义或可称为“动”或“变”，取自“以动者尚其变”。
